File: designing and implementing/design_stack.py
# designing stack using linked list
class Node:
    def __init__(self, data):
       self.data = data
       self.next = None

class Stack:
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0
    
    def peek():
        return self.top.value
    
    def push(value):
        newNode = Node(value)
        if length == 0:
            self.top = newNode
            self.bottom = newNode
        else:
            holdinPointers = self.top
            self.top = newNode
            self.top.next = holdinPointers
        
        self.length += 1
        
    def pop():
        if not self.top:
            return None
        
        if self.top == self.bottom:
            self.bottom = None

        self.top = self.top.next
        self.length -= 1
        
        
        

File: design and implementing/design_stack.js
// designing stack using linked list

class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.bottom = null;
    this.top = null;
    this.length = 0;
  }

  peek() {
    return this.top;
  }

  push(value) {
    const newNode = new Node(value);

    if (this.length === 0) {
      this.top = newNode;
      this.bottom = newNode;
    } else {
      const holdingPointers = this.top;
      this.top = newNode;
      this.top.next = holdingPointers;
    }

    this.length += 1;
  }

  pop() {
    if (this.lenght === 0) {
      return null;
    }

    if (this.top === this.bottom) {
      this.bottom = none;
    }

    this.top = this.top.next;

    this.length -= 1;
  }
}
